torch/distributed/_functional_collectives.py

```diff
@@ -363,26 +363,52 @@ def reduce_scatter_tensor(
         tensor_list = torch.chunk(self, group_size, dim=scatter_dim)
         self = torch.cat(tensor_list)
 
-    tensor = torch._C._nn.reduce_scatter_tensor(self, reduceOp, 0, tag, rankset, group_size)  # type: ignore[attr-defined]
+    tensor = torch.ops.c10d_functional.reduce_scatter_tensor(self, reduceOp, 0, tag, rankset, group_size)  # type: ignore[attr-defined]
     res = _maybe_wrap_tensor(tensor)
     return res
 
-
-c10_lib_cpu = torch.library.Library("aten", "IMPL", "CPU")
-c10_lib_cuda = torch.library.Library("aten", "IMPL", "CUDA")
+c10_lib = torch.library.Library("c10d_functional", "DEF")
+c10_lib_impl = torch.library.Library("c10d_functional", "IMPL")
 
 def _register_ops():
-    c10_lib_cpu.impl("all_reduce", _all_reduce)
-    c10_lib_cuda.impl("all_reduce", _all_reduce)
+    c10_lib.define("all_reduce(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor")
+    c10_lib.define("wait_tensor(Tensor self) -> Tensor")
+    c10_lib.define("all_gather_into_tensor(Tensor shard, str tag, int[] ranks, int group_size) -> Tensor")
+    c10_lib.define(
+        "reduce_scatter_tensor(Tensor input, str reduceOp, int scatter_dim, str tag, int[] ranks, int group_size) -> Tensor"
+    )
+
+    c10_lib_impl.impl("all_reduce", _all_reduce, "CPU")
+    c10_lib_impl.impl("all_reduce", _all_reduce, "CUDA")
+
+    c10_lib_impl.impl("wait_tensor", _wait_tensor, "CPU")
+    c10_lib_impl.impl("wait_tensor", _wait_tensor, "CUDA")
+
+    c10_lib_impl.impl("all_gather_into_tensor", _all_gather_into_tensor, "CPU")
+    c10_lib_impl.impl("all_gather_into_tensor", _all_gather_into_tensor, "CUDA")
+
+    c10_lib_impl.impl("reduce_scatter_tensor", _reduce_scatter_tensor, "CPU")
+    c10_lib_impl.impl("reduce_scatter_tensor", _reduce_scatter_tensor, "CUDA")
+
+    # We now register meta kernels to deal with tracing
+    def empty_like_self(self, *args):
+        return torch.empty_like(self)
+    c10_lib_impl.impl("all_reduce", empty_like_self, "Meta")
+    c10_lib_impl.impl("wait_tensor", empty_like_self, "Meta")
+
+    def all_gather_into_tensor_meta(shard, tag, rankset, group_size):
+        out_size = list(shard.size())
+        out_size[0] *= group_size
+        return shard.new_empty(out_size)
+    c10_lib_impl.impl("all_gather_into_tensor", all_gather_into_tensor_meta, "Meta")
```

Comment:
@albanD whats the best way to achieve
 - specific meta registration per op
 - support for each op across many devices, ideally without per-specific-device keys here
 
 ?
 
 Rodrigo tried using BackendSelect for the op, but found it also made the meta key hit those actual ops instead of their meta kernels.

Reply:
It sounds like you want the op to be a primitive in the graph, but to have a generic implementation that works across devices? I think you want to register the main impl to the `CompositeExplicitAutograd` key.

Reply:
> specific meta registration per op

Exactly as done in this line!

> support for each op across many devices, ideally without per-specific-device keys here

If you have a generic impl that works for all device, you can register it to "CompositeExplicitAutograd" key. Or if it is only a subset of the devices, you can register to the right keys: "CPU", "CUDA", "Lazy", etc.

Reply:
yea, i think 
* for tracing backends, we just really want composite so we can trace generically and rely on backends to either work or err on missing support
* for eager backends, we call into c10d already, and it provides indirection via processgroup.  Maybe its enough indirection that we can also get away with raising the error in the backend if a device isn't supported

so modulo if i misunderstand how things work for eager, i think registering for CompositeExplicitAutograd makes sense here (and we should put in a comment about our assumptions too)